### PR TITLE
ntuple production for hadronic resolved top tagging

### PR DESCRIPTION
**Pull request by matt-komm**

### CODE REVIEW

File: processors/ttX3.py [hunk 1]
```diff
@@ -240,7 +240,6 @@ def jetSelection(jetDict):
 analyzerChain = [
     EventSkim(selection=lambda event: event.nTrigObj > 0),
     MetFilter(
-        globalOptions=globalOptions,
```
Comment (demuller) [inline processors/ttX3.py:243]: Any reason why the global options are omitted here?

File: python/modules/MetFilter.py [hunk 2]
```diff
@@ -9,8 +9,7 @@
 from PhysicsTools.NanoAODTools.postprocessing.framework.eventloop import Module
 
 class MetFilter(Module):
-    def __init__(self,globalOptions={"isData":False}, outputName=None):
-        self.globalOptions=globalOptions
+    def __init__(self,outputName=None):
```
Comment (demuller) [inline python/modules/MetFilter.py:12]: see comment on global options above

File: python/modules/TopNNReco.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+            'jet_neHEF': lambda event,lepton,jet: jet.neHEF,
+            
+            'jet_pt': lambda event,lepton,jet: jet.pt,
+            'jet_abseta': lambda event,lepton,jet: math.fabs(jet.eta),
+            'jet_mass': lambda event,lepton,jet: jet.mass,
+            
+            
+            #lepton vars
+            'lepton_pt': lambda event,lepton,jet: lepton.pt,
+            'lepton_abseta': lambda event,lepton,jet: math.fabs(lepton.pt),
```
Comment (demuller) [inline python/modules/TopNNReco.py]: `lepton.pt` --> `lepton.eta`

File: python/modules/TopNNReco.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+            
+            'lepton_dxy': lambda event,lepton,jet:lepton.dxy,
+            'lepton_dxysig': lambda event,lepton,jet:lepton.dxy/(1e-4+math.fabs(lepton.dxyErr)),
+            'lepton_dz': lambda event,lepton,jet:lepton.dxy,
```
Comment (demuller) [inline python/modules/TopNNReco.py]: `lepton.dxy` --> `lepton.dz`

File: python/modules/TopNNReco.py [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+            'lepton_dzsig': lambda event,lepton,jet:lepton.dz/(1e-4+math.fabs(lepton.dzErr)),
+            
+            'lepton_isMuon': lambda event,lepton,jet:1. if lepton.isMuon else 0.,
+            
+            
+            #electron specific vars
+            'ele_deltaEtaSC': lambda event,lepton,jet:math.log(max(1e-5,math.fabs(lepton.deltaEtaSC))) if lepton.isElectron else 0.,
```
Comment (demuller) [inline python/modules/TopNNReco.py:57]: Why log?

File: python/modules/TopNNReco.py [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+            'ele_dr03EcalRecHitSumEt': lambda event,lepton,jet:lepton.dr03EcalRecHitSumEt if lepton.isElectron else 0.,
+            'ele_dr03HcalDepth1TowerSumEt': lambda event,lepton,jet:lepton.dr03HcalDepth1TowerSumEt if lepton.isElectron else 0.,
+            'ele_dr03TkSumPt': lambda event,lepton,jet:lepton.dr03TkSumPt if lepton.isElectron else 0.,
+            'ele_dr03TkSumPtHEEP': lambda event,lepton,jet:lepton.dr03TkSumPtHEEP if lepton.isElectron else 0.,
+            
+            'ele_eInvMinusPInv': lambda event,lepton,jet:lepton.eInvMinusPInv if lepton.isElectron else 0.,
+            'ele_hoe': lambda event,lepton,jet:lepton.hoe if lepton.isElectron else 0.,
+            'ele_ip3d': lambda event,lepton,jet:lepton.ip3d if lepton.isElectron else 0.,
+            'ele_jetPtRelv2': lambda event,lepton,jet: math.log(max(1e-5,lepton.jetPtRelv2)) if lepton.isElectron else 0.,
```
Comment (demuller) [inline python/modules/TopNNReco.py:66]: Why log?

File: python/modules/TopNNReco.py [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+            'ele_jetRelIso': lambda event,lepton,jet:math.log(max(1e-5,lepton.jetRelIso)) if lepton.isElectron else 0.,
```
Comment (demuller) [inline python/modules/TopNNReco.py:67]: Why log? (for all variables that use log)

File: python/modules/TopNNReco.py [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+                leptonicTops.append(top)
+            else:
+                # rare more complex decays (e.g. involving taus or FSR)
+                otherTops.append(top)
+                
+        return {"hadronic": hadronicTops, "leptonic": leptonicTops, "other": otherTops}
+
+    def makeLeptonJetCombinations(self, leptons, jets):
+        combinations = []
+        for lepton in leptons:
```
Comment (demuller) [inline python/modules/TopNNReco.py:374]: Consider using `from itertools import permutations` to obtain all jet lepton permutations:
```
perm = permutations(leptons,jets)
```

File: python/modules/TopNNReco.py [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+                combinations.append([lepton,jet])
+        return combinations
+        
+    def makeJetCombinations(self, bjets, jets):
+        combinations = []
+        
+        # sort by pT so that pt(jet1)>pt(jet2)
+        jets = sorted(jets,key=lambda x: x.pt,reverse=True)
+        
+        for bjet in bjets:
```
Comment (demuller) [inline python/modules/TopNNReco.py:386]: same comment as for lepton jet combinations

**Review by demuller — CHANGES_REQUESTED** [review]
Hi @matt-komm I found 2-3 minor errors and have also some questions/suggestions. Please also add a general description to the PR

File: processors/ttX3.py (hunk 1 above)
Comment (matt-komm) [reply processors/ttX3.py:243]: this is now a static variable

File: python/modules/TopNNReco.py (hunk 8 above)
Comment (matt-komm) [reply python/modules/TopNNReco.py:374]: better to allow for additional filtering in the loop

**Comment by matt-komm** [comment]
"ntuple production for hadronic resolved top tagging" seems a good enough title and description for me :-)

**Pull request merged**